### PR TITLE
fix: validate agent & resource metadata keys during plan

### PR DESCRIPTION
**Pull request by ethanndickson**
Closes https://github.com/coder/coder/issues/14387.
Closes https://github.com/coder/coder/issues/14569.

Previously, the configuration had to be applied in order for the provider to complain about duplicate keys - now just a plan will suffice.

### CODE REVIEW

**Comment by ethanndickson** [comment]
* **#277** <a href="https://app.graphite.dev/github/pr/coder/terraform-provider-coder/277?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ethanndickson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: provider/agent.go [hunk 1]
```diff
@@ -272,6 +260,23 @@ func agentResource() *schema.Resource {
 				Optional:    true,
 			},
 		},
+		CustomizeDiff: func(ctx context.Context, rd *schema.ResourceDiff, i interface{}) error {
+			if !rd.HasChange("metadata") {
+				return nil
+			}
+
+			keys := map[string]struct{}{}
```
Comment (johnstcn) [inline provider/agent.go:268]: Nit: If you change this to a `map[string]bool` you can make the below condition look a lot nicer, at the expense of a few bytes:

```
if keys[key] {
  return ...
}
```

File: provider/agent.go [hunk 2]
```diff
@@ -272,6 +260,23 @@ func agentResource() *schema.Resource {
...
+			for _, t := range rd.Get("metadata").([]interface{}) {
```
Comment (johnstcn) [inline provider/agent.go:269]: Would it make sense to check that `rd.Get("metadata")` can be cast to a `[]interface{}`?
Same for below with `taint`.

Also, I think you can just use `any` as an alias for `interface{}` now.

File: provider/metadata.go [hunk 3]
```diff
@@ -111,5 +112,22 @@ func metadataResource() *schema.Resource {
 				},
 			},
 		},
+		CustomizeDiff: func(ctx context.Context, rd *schema.ResourceDiff, i interface{}) error {
+			if !rd.HasChange("item") {
+				return nil
+			}
+
+			keys := map[string]struct{}{}
+			for _, t := range rd.Get("item").([]interface{}) {
```
Comment (johnstcn) [inline provider/metadata.go:121]: check cast?

File: integration/integration_test.go [hunk 4]
```diff
@@ -103,7 +103,7 @@ func TestIntegration(t *testing.T) {
 				"workspace.transition":              `start`,
 				"workspace_owner.email":             `testing@coder\.com`,
 				"workspace_owner.full_name":         `default`,
-				"workspace_owner.groups":            `\[\]`,
+				"workspace_owner.groups":            `\[\"Everyone\"\]`,
```
Comment (johnstcn) [inline integration/integration_test.go:106]: Integration test doesn't appear to like this. Should this be moved to a separate PR?

File: provider/agent.go (hunk 2 above)
Comment (ethanndickson) [reply provider/agent.go:269]: I don't think it's possible for either cast to fail, but I'll add checks in-case the schema changes

File: integration/integration_test.go (hunk 4 above)
Comment (ethanndickson) [reply integration/integration_test.go:106]: That's annoying, I thought this would appease.. I'll make a new PR.

**Review by johnstcn — APPROVED** [review]

**Comment by matifali** [comment]
Do we want to port this to `release/1.0` and do a patch?

**Comment by ethanndickson** [comment]
Yeah, potentially. What are the breaking changes on main not included in any release?

**Comment by matifali** [comment]
We removed a couple of deprecated resources and attributes.
